src/fastcs/mapping.py

```diff
@@ -0,0 +1,51 @@
+from dataclasses import dataclass
+from typing import Callable, NamedTuple
+
+from .attributes import Attribute
+from .controller import BaseController, Controller
+from .cs_methods import MethodInfo
+
+MethodData = NamedTuple(
+    "MethodData", (("name", str), ("info", MethodInfo), ("method", Callable))
+)
+
+
+@dataclass
+class SingleMapping:
+    controller: BaseController
+    methods: list[MethodData]
+    attributes: dict[str, Attribute]
+
+
+class Mapping:
+    def __init__(self, controller: Controller) -> None:
+        self.controller = controller
+        self._generate_mapping(controller)
+
+    @staticmethod
+    def _get_single_mapping(controller: BaseController) -> SingleMapping:
+        methods = []
+        attributes = {}
+        for attr_name in dir(controller):
+            attr = getattr(controller, attr_name)
+            if hasattr(attr, "fastcs_method_info"):
+                methods.append(MethodData(attr_name, attr.fastcs_method_info, attr))
+            elif isinstance(attr, Attribute):
+                attributes[attr_name] = attr
+
+        return SingleMapping(controller, methods, attributes)
```

Comment:
Could this logic be part of the `Controller` and called on `__init__` to populate `cs_attributes` and `cs_methods` members of itself? Then in the backends it could just interrogate the `Controller` instance itself to pull out the relevant attributes and methods and we don't need the `Mapping` or `SingleMapping` classes.

Reply:
I would prefer to wait until we've tested a programmatic interface before we do something like that.

The other concern is that the Controller is getting used directly by the end user, while the Mapping is used by the Backend classes, which we will be developing ourselves. Merging the two public interfaces will overcomplicate the documentation etc for the end user.

I think a reorganisation (and renaming!) of Mapping would be a better idea.